packages/react/src/ReferenceInput/ReferenceInput.tsx

```diff
@@ -16,75 +17,253 @@ export interface ReferenceInputProps {
+  value: string;
+}
+
+type ProfileTargetType = BaseTargetType & {
+  type: 'profile';
+  name?: string;
+  title?: string;
+  resourceType?: string;
+  error?: any;
+};
+
+type ResourceTypeTargetType = BaseTargetType & {
+  type: 'resourceType';
+  resourceType: string;
+};
+type TargetType = ResourceTypeTargetType | ProfileTargetType;
+
 export function ReferenceInput(props: ReferenceInputProps): JSX.Element {
-  const targetTypes = getTargetTypes(props.targetTypes);
-  const initialResourceType = getInitialResourceType(props.defaultValue, targetTypes);
+  const medplum = useMedplum();
   const [value, setValue] = useState<Reference | undefined>(props.defaultValue);
-  const [resourceType, setResourceType] = useState<ResourceType | undefined>(initialResourceType);
+  const [targetTypes, setTargetTypes] = useState<TargetType[] | undefined>(() => createTargetTypes(props.targetTypes));
+  const [state, setState] = useState<{
+    targetType?: TargetType;
+    resourceType?: ResourceType;
+  }>(() => {
+    const targetType = getInitialTargetType(props.defaultValue, targetTypes);
+    return { targetType, resourceType: targetType?.resourceType as ResourceType };
+  });
 
-  function setValueHelper(newValue: Reference | undefined): void {
-    setValue(newValue);
-    if (props.onChange) {
-      props.onChange(newValue);
+  const searchCriteria = useMemo<ReferenceInputProps['searchCriteria']>(() => {
+    if (state.targetType?.type === 'profile') {
+      return { ...props.searchCriteria, _profile: state.targetType.value };
     }
-  }
+    return props.searchCriteria;
+  }, [props.searchCriteria, state.targetType]);
+
+  useEffect(() => {
+    if (!targetTypes) {
+      return;
+    }
+
+    if (targetTypes.filter(shouldFetchResourceType).length === 0) {
+      return;
+    }
+
+    const newTargetTypePromises: Promise<TargetType>[] = targetTypes.map((tt) => {
+      if (shouldFetchResourceType(tt)) {
+        return fetchResourceTypeOfProfile(medplum, tt.value)
+          .then((profile) => {
+            const newTargetType = { ...tt };
+
+            if (!profile) {
+              console.debug(`StructureDefinition for ${tt.value} not found`);
+              newTargetType.error = 'StructureDefinition not found';
+            } else if (!profile.type || isEmpty(profile.type)) {
+              console.debug(`resourceType for ${tt.value} unexpectedly missing`);
+              newTargetType.error = 'resourceType unavailable';
+            } else {
+              newTargetType.resourceType = profile.type satisfies string;
+              newTargetType.name = profile.name;
+              newTargetType.title = profile.title;
+            }
+
+            return newTargetType;
+          })
+          .catch((reason) => {
+            console.error(reason);
+            return { ...tt, error: reason };
+          });
+      } else {
+        return Promise.resolve(tt);
+      }
+    });
+
+    Promise.all(newTargetTypePromises)
```

Comment:
Did you test it with multiple profiles?  Auto-batching should pick it up and combine it into a single HTTP call.  "should"

Reply:
They did not auto-batch since GQL requests use `POST` instead of `GET`.

Reply:
Got it.  Sorry, I should have known that, sorry for the goose chase!